terraform/transform_provider.go

```diff
@@ -360,18 +355,28 @@ func (t *MissingProviderTransformer) Transform(g *Graph) error {
 		// it's safe for us to rely on ProvidedBy here rather than waiting for
 		// the later proper resolution of provider inheritance done by
 		// ProviderTransformer.
-		p, _ := pv.ProvidedBy()
+		providerAddr, _ := pv.ProvidedBy()
 		var providerFqn addrs.Provider
-		switch p.(type) {
+		switch p := providerAddr.(type) {
```

Comment:
It looks like we have a few different quite-similar switch statements like this now, though they are not identical. Do you think it would be practical/useful (likely in a separate PR) to write a helper function that takes an `addrs.ProviderConfig`, the node's implied provider, and a config and then encapsulates this logic for selecting an appropriate `addrs.Provider`?

I'm imagining a function with a signature like this:

```go
func providerFQNForProviderConfig(addr addrs.ProviderConfig, rootConfig *configs.Config, impliedAddr addrs.Provider) addrs.Provider
```

It looks like there would still be some one-off logic in the callers of this hypothetical `providerFQNForProviderConfig`, like the special case of ignoring aliased provider configurations in _this_ caller, but at first glance it seems like those things are separable from the main switch statement so that we could factor this out into an easily-testable unit. :thinking:

Reply:
I agree that something like this would be handy, though I'd make the following changes (these are more notes to myself):

* passing both an addrs.ProviderConfig and addrs.Provider is redundant - we already have a function to look up FQN via localconfig so this would only be useful when you only have an addrs.Provider
* either replace `rootConfig` with the module config or also pass the module path, since the mapping tables are module-specific